### ISSUE TITLE
UPDATE .. RETURNING cannot be used Reactively

### ISSUE THREAD

**Issue by chris-dekker** [jOOQ/jOOQ#11190]
### Expected behavior

Whenever I use UPDATE .. RETURNING I expect the return type to be a publisher of the result set requested in the `returning()` call.

### Actual behavior

Whenever I try to use UPDATE .. RETURNING to return the updated record in a reactive context I am unable to get the publisher for the returning result. Only the number of updated records is returned as a publisher.

### Steps to reproduce the problem

Consider the following examples:

```
Mono.from(
	context
		.selectFrom(USERS)
		.where(USERS.ID.eq(123L)))
// returns Mono<UserRecord>
		
Mono.from(
	context
		.update(USERS)
		.set(USERS.NAME, "Chris")
		.where(USERS.ID.eq(123L)))
// returns Mono<Integer> with number of records updated

Mono.from(
	context
		.update(USERS)
		.set(USERS.NAME, "Chris")
		.where(USERS.ID.eq(123L))
		.returning())
// still returns Mono<Integer> with number of records updated (!!)
```
The first select returns a `SelectConditionStep<R>` which extends all the way up to a `ResultQuery<R>`, which extends `org.reactivestreams.Publisher<R>`. (where `R extends Record`)

The update returns a `UpdateConditionStep<R>` which extends all the way up to a non-generic `RowCountQuery` which extends `org.reactivestreams.Publisher<Integer>`.

So far so good, but once you add `returning()` it still returns a `UpdateResultStep<R>` which extends the same `RowCountQuery`, losing access to a relevant publisher of type `R`.

**When using `returning(..)` the return type should implement a `org.reactivestreams.Publisher<R>` somewhere in the hierarchy instead of the `org.reactivestreams.Publisher<Integer>` tied to the update queries.**

### Versions

- jOOQ: 3.14.4
- Java: 11.0.8
- Database (include vendor): PostgreSQL
- OS: N/A
- JDBC Driver (include name if inofficial driver): N/A

**Comment by lukaseder** [comment]
Thanks a lot for your message.

Yes, this is an unfortunate limitation. We'd have to return a type incompatible with `RowCountQuery` on any `returning()` call, but that cannot be done backwards compatibly. For this reason, a new type had to be introduced to solve some other issues in the sync world, namely to be able to use all the other `fetch()` methods from `ResultQuery` on a new `Fetchable` type: https://github.com/jOOQ/jOOQ/issues/3185

The `Fetchable` type could produce a `Publisher<R>` via a new `fetchingPublisher()` method (or similar).

**Comment by lukaseder** [comment]
I'm currently looking into this in the context of https://github.com/jOOQ/jOOQ/issues/11700

**Comment by lukaseder** [comment]
It works well with a `Fetchable.publisher()` method, which would offer the feature backwards compatibly. But I'll try to see if we can pull of a "reasonable" backwards incompatible change. Once someone calls `returning()`, the update count is hardly of interest, nor is a synchronous call to `execute()`. Users will *always* want the `Result<R>` or `Publisher<R>` instead.

If we don't fix this, then depending on the complexity of type inference in a reactor or RxJava client call chain, the compiler errors might be quite cryptic when users expect `Something<R>` but instead are getting a `Integer`

**Comment by lukaseder** [comment]
The backwards incompatibility this would introduce is that 

- `XReturningStep<R>` would continue to extend `XFinalStep<R> <: X<R> <: RowCountQuery <: Query`
- `XReturningStep<R>.returning()` would continue to return `XResultStep<R>`
- `XResultStep<R>` would no longer extend `X<R>`, but `Query & Fetchable<R>`, where `Fetchable<R> <: Iterable<R> & Publisher<R>`

In the above list, `X<R>` is `Insert<R> | Update<R> | Delete<R>`. Since those are mostly useless marker interfaces, I wonder if we could just deprecate them, similar to all the many `XYZFinalStep` types (https://github.com/jOOQ/jOOQ/issues/11329)

**Comment by lukaseder** [comment]
No, deprecation is too harsh, and may not be necessary...

**Comment by lukaseder** [comment]
Related, new issue: https://github.com/jOOQ/jOOQ/issues/11720

**Comment by lukaseder** [comment]
This will be fixed in jOOQ 3.15 via #11720. In summary, we'll introduce a breaking change. As soon as `returning()` is called on any of the `INSERT`, `UPDATE`, or `DELETE` DSL, the resulting type is no longer a `RowCountQuery` but a `ResultQuery`
